File: client/js/crypto.js
/**
 * Layer 1: X25519 key exchange + AES-256-GCM encryption.
 *
 * Handles identity key generation, session key derivation via HKDF,
 * directional nonce construction, and anti-replay protection.
 */

import { b64Encode, b64Decode, concatBuffers } from './utils.js';

export { b64Encode, b64Decode };

const REPLAY_WINDOW = 64;

export class RelayCrypto {
  constructor() {
    this.keyPair = null;
    this.sessionKey = null;
    this.sendCounter = 0;
    this.recvCounterMax = -1;
    this.recvWindow = new Set();
    this.clientNonce = null;
    this.publicKeyBytes = null;
  }

  async generateKeyPair() {
    this.keyPair = await crypto.subtle.generateKey(
      { name: 'X25519' }, true, ['deriveBits']
    );
    const rawPub = await crypto.subtle.exportKey('raw', this.keyPair.publicKey);
    this.publicKeyBytes = new Uint8Array(rawPub);
    this.clientNonce = crypto.getRandomValues(new Uint8Array(32));
  }

  /**
   * Regenerate only the session nonce (for reconnections with the same identity keypair).
   * The keypair stays the same; a fresh nonce ensures a unique session key.
   */
  regenerateNonce() {
    this.clientNonce = crypto.getRandomValues(new Uint8Array(32));
  }

  async deriveSessionKey(gatewayPubKeyBytes, gatewayNonce) {
    // Import the gateway's public key
    const gatewayPubKey = await crypto.subtle.importKey(
      'raw', gatewayPubKeyBytes, { name: 'X25519' }, true, []
    );

    // X25519 ECDH
    const sharedSecret = await crypto.subtle.deriveBits(
      { name: 'X25519', public: gatewayPubKey },
      this.keyPair.privateKey, 256
    );

    // salt = SHA-256(clientPub || gatewayPub || clientNonce || gatewayNonce)
    const saltInput = concatBuffers(
      this.publicKeyBytes,
      gatewayPubKeyBytes,
      this.clientNonce,
      gatewayNonce
    );
    const salt = await crypto.subtle.digest('SHA-256', saltInput);

    // HKDF key derivation
    const hkdfKey = await crypto.subtle.importKey(
      'raw', sharedSecret, 'HKDF', false, ['deriveKey']
    );
    this.sessionKey = await crypto.subtle.deriveKey(
      {
        name: 'HKDF',
        hash: 'SHA-256',
        salt: new Uint8Array(salt),
        info: new TextEncoder().encode('openclaw-relay-v1')
      },
      hkdfKey,
      { name: 'AES-GCM', length: 256 },
      false,
      ['encrypt', 'decrypt']
    );

    this.sendCounter = 0;
    this.recvCounterMax = -1;
    this.recvWindow.clear();
  }

  async encrypt(plaintext) {
    const nonce = new Uint8Array(12);
    const dv = new DataView(nonce.buffer);
    dv.setUint32(0, 1); // direction: client -> gateway
    const counter = this.sendCounter++;
    const high = Math.floor(counter / 0x100000000);
    const low = counter >>> 0;
    dv.setUint32(4, high);
    dv.setUint32(8, low);

    const ciphertext = await crypto.subtle.encrypt(
      { name: 'AES-GCM', iv: nonce },
      this.sessionKey,
      new TextEncoder().encode(plaintext)
    );

    // nonce(12) || ciphertext+tag
    return b64Encode(concatBuffers(nonce, new Uint8Array(ciphertext)));
  }

  async decrypt(payload) {
    const raw = b64Decode(payload);
    if (raw.length < 12 + 16) {
      throw new Error('Ciphertext too short');
    }

    const nonce = raw.slice(0, 12);
    const ciphertextAndTag = raw.slice(12);

    // Validate direction prefix: client must receive direction 2 (gateway->client)
    const dv = new DataView(nonce.buffer, nonce.byteOffset, 12);
    const direction = dv.getUint32(0);
    if (direction !== 2) {
      throw new Error(`Wrong nonce direction: expected 2 (gw->client), got ${direction}`);
    }

    // Extract counter from nonce bytes 4-11
    const high = dv.getUint32(4);
    const low = dv.getUint32(8);
    const counter = high * 0x100000000 + low;

    // Anti-replay check
    if (this.recvCounterMax < 0) {
      if (counter !== 0) {
        throw new Error('Replay detected: first counter must be zero');
      }
    } else {
      if (counter <= this.recvCounterMax - REPLAY_WINDOW) {
        throw new Error('Replay detected: counter too old');
      }
      if (counter <= this.recvCounterMax && this.recvWindow.has(counter)) {
        throw new Error('Replay detected: duplicate counter');
      }
    }

    const plaintext = await crypto.subtle.decrypt(
      { name: 'AES-GCM', iv: nonce },
      this.sessionKey,
      ciphertextAndTag
    );

    // Update replay bookkeeping after successful decryption
    if (counter > this.recvCounterMax) {
      this.recvCounterMax = counter;
    }
    this.recvWindow.add(counter);
    const cutoff = this.recvCounterMax - REPLAY_WINDOW;
    for (const c of this.recvWindow) {
      if (c <= cutoff) this.recvWindow.delete(c);
    }

    return new TextDecoder().decode(plaintext);
  }
}

/**
 * Build a 12-byte GCM nonce: [4-byte direction][8-byte counter].
 * Exported for testing.
 */
export function buildNonce(direction, counter) {
  const nonce = new Uint8Array(12);
  const dv = new DataView(nonce.buffer);
  dv.setUint32(0, direction);
  dv.setUint32(4, Math.floor(counter / 0x100000000));
  dv.setUint32(8, counter >>> 0);
  return nonce;
}

/**
 * Check whether a counter passes the anti-replay window.
 * Returns 'ok', 'too_old', or 'duplicate'.
 * Exported for testing.
 */
export function checkReplay(counter, recvCounterMax, recvWindow) {
  if (recvCounterMax < 0) {
    return counter === 0 ? 'ok' : 'invalid_first_counter';
  }
  if (counter <= recvCounterMax - REPLAY_WINDOW) return 'too_old';
  if (counter <= recvCounterMax && recvWindow.has(counter)) return 'duplicate';
  return 'ok';
}

File: sdk/python/openclaw_relay/crypto.py
"""Layer 1: X25519 key exchange, AES-256-GCM encryption, HKDF key derivation."""

from __future__ import annotations

import hashlib
import os
import struct

from cryptography.hazmat.primitives.asymmetric.x25519 import (
    X25519PrivateKey,
    X25519PublicKey,
)
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.hashes import SHA256
from cryptography.hazmat.primitives.kdf.hkdf import HKDF

_HKDF_INFO = b"openclaw-relay-v1"
_REPLAY_WINDOW_SIZE = 64


class KeyPair:
    """X25519 keypair for Diffie-Hellman key exchange."""

    def __init__(self) -> None:
        self.private_key = X25519PrivateKey.generate()
        self.public_key = self.private_key.public_key()

    @property
    def public_key_bytes(self) -> bytes:
        """Return the raw 32-byte public key."""
        from cryptography.hazmat.primitives.serialization import (
            Encoding,
            PublicFormat,
        )

        return self.public_key.public_bytes(Encoding.Raw, PublicFormat.Raw)

    @classmethod
    def from_private_bytes(cls, data: bytes) -> KeyPair:
        """Reconstruct a KeyPair from raw 32-byte private key material."""
        instance = object.__new__(cls)
        instance.private_key = X25519PrivateKey.from_private_bytes(data)
        instance.public_key = instance.private_key.public_key()
        return instance


def compute_shared_secret(
    private_key: X25519PrivateKey,
    peer_public_key_bytes: bytes,
) -> bytes:
    """Perform X25519 ECDH and return the 32-byte shared secret."""
    peer_public_key = X25519PublicKey.from_public_bytes(peer_public_key_bytes)
    return private_key.exchange(peer_public_key)


def derive_session_key(
    shared_secret: bytes,
    client_public_key: bytes,
    gateway_public_key: bytes,
    client_session_nonce: bytes,
    gateway_session_nonce: bytes,
) -> bytes:
    """Derive a 32-byte AES-256 session key using HKDF-SHA256.

    salt = SHA256(client_pub || gateway_pub || client_nonce || gateway_nonce)
    key  = HKDF(ikm=shared_secret, salt=salt, info="openclaw-relay-v1", length=32)
    """
    salt_input = client_public_key + gateway_public_key + client_session_nonce + gateway_session_nonce
    salt = hashlib.sha256(salt_input).digest()

    hkdf = HKDF(
        algorithm=SHA256(),
        length=32,
        salt=salt,
        info=_HKDF_INFO,
    )
    return hkdf.derive(shared_secret)


def generate_session_nonce() -> bytes:
    """Generate a random 32-byte session nonce."""
    return os.urandom(32)


def channel_token_hash(token: str) -> str:
    """Return the SHA-256 hex digest of *token*."""
    return hashlib.sha256(token.encode()).hexdigest()


class SessionCipher:
    """AES-256-GCM encryption / decryption with monotonic nonce counters.

    The 12-byte GCM nonce is structured as:
        [4 bytes direction prefix (big-endian)] [8 bytes counter (big-endian)]

    A sliding-window replay check (window size 64) prevents duplicate
    or out-of-order replay of received messages.
    """

    DIRECTION_CLIENT_TO_GATEWAY = 1
    DIRECTION_GATEWAY_TO_CLIENT = 2

    def __init__(self, session_key: bytes, send_direction: int) -> None:
        if len(session_key) != 32:
            raise ValueError("session_key must be 32 bytes")
        if send_direction not in (
            self.DIRECTION_CLIENT_TO_GATEWAY,
            self.DIRECTION_GATEWAY_TO_CLIENT,
        ):
            raise ValueError("send_direction must be 1 (client->gw) or 2 (gw->client)")

        self._aesgcm = AESGCM(session_key)
        self._send_direction = send_direction
        self._recv_direction = (
            self.DIRECTION_GATEWAY_TO_CLIENT
            if send_direction == self.DIRECTION_CLIENT_TO_GATEWAY
            else self.DIRECTION_CLIENT_TO_GATEWAY
        )
        self._send_counter: int = 0

        # Receive-side replay protection
        self._recv_counter: int = 0  # highest counter seen so far
        self._recv_window: set[int] = set()

    # ------------------------------------------------------------------
    # Encryption
    # ------------------------------------------------------------------

    def encrypt(self, plaintext: bytes) -> bytes:
        """Encrypt *plaintext* and return ``nonce(12) || ciphertext || tag(16)``."""
        nonce = struct.pack(">I", self._send_direction) + struct.pack(">Q", self._send_counter)
        self._send_counter += 1
        ct = self._aesgcm.encrypt(nonce, plaintext, None)
        return nonce + ct

    # ------------------------------------------------------------------
    # Decryption
    # ------------------------------------------------------------------

    def decrypt(self, data: bytes) -> bytes:
        """Decrypt *data* (``nonce(12) || ciphertext || tag(16)``).

        Raises ``ValueError`` on replay or authentication failure.
        """
        if len(data) < 12 + 16:
            raise ValueError("Ciphertext too short")

        nonce = data[:12]
        ciphertext_and_tag = data[12:]

        # Validate direction prefix
        direction = struct.unpack(">I", nonce[:4])[0]
        if direction != self._recv_direction:
            raise ValueError(
                f"Wrong nonce direction prefix: expected {self._recv_direction}, got {direction}"
            )

        counter = struct.unpack(">Q", nonce[4:12])[0]

        # Anti-replay ---------------------------------------------------
        if self._recv_counter > 0 or len(self._recv_window) > 0:
            if counter <= self._recv_counter - _REPLAY_WINDOW_SIZE:
                raise ValueError("Replay detected: counter too old")
            if counter <= self._recv_counter and counter in self._recv_window:
                raise ValueError("Replay detected: duplicate counter")

        plaintext = self._aesgcm.decrypt(nonce, ciphertext_and_tag, None)

        # Update bookkeeping only after successful decryption
        if counter > self._recv_counter:
            self._recv_counter = counter
        self._recv_window.add(counter)

        # Trim the window to keep only the last 64 counters
        self._recv_window = {c for c in self._recv_window if c > self._recv_counter - _REPLAY_WINDOW_SIZE}

        return plaintext
